File: 164-maximum-gap/maximum-gap.cpp
class Solution {
public:
    int maximumGap(vector<int>& nums) {
        int n = nums.size();
        if (n < 2)
            return 0;

        int mini = *min_element(nums.begin(), nums.end());
        int maxi = *max_element(nums.begin(), nums.end());

        // bucket size (minimum possible max gap)
        int bucketSize = ceil((double)(maxi - mini) / (n - 1));
        if (bucketSize == 0)
            bucketSize = 1;

        // number of buckets = n - 1
        vector<int> minBucket(n - 1, INT_MAX);
        vector<int> maxBucket(n - 1, INT_MIN);

        // distribute numbers into buckets
        for (int num : nums) {
            if (num == mini || num == maxi)
                continue;
            int idx = (num - mini) / bucketSize;
            minBucket[idx] = min(minBucket[idx], num);
            maxBucket[idx] = max(maxBucket[idx], num);
        }

        // compute maximum gap
        int ans = 0;
        int prev = mini;

        for (int i = 0; i < n - 1; i++) {
            if (minBucket[i] == INT_MAX)
                continue; // empty bucket
            ans = max(ans, minBucket[i] - prev);
            prev = maxBucket[i];
        }

        // final gap with maximum element
        ans = max(ans, maxi - prev);

        return ans;
    }
};

File: 164-maximum-gap/maximum-gap.java
class Solution {
    public int maximumGap(int[] nums) {
        int n = nums.length;
        if (n == 1)
            return 0;
        int mini = nums[0], maxi = nums[0];
        for (int num : nums) {
            mini = Math.min(mini, num);
            maxi = Math.max(maxi, num);
        }

        int bucketSize = ((maxi - mini) + (n - 1) - 1) / (n - 1);
        int bucketLength = n - 1;
        int[] bucketMin = new int[bucketLength];
        Arrays.fill(bucketMin, Integer.MAX_VALUE);
        int[] bucketMax = new int[bucketLength];
        Arrays.fill(bucketMax, Integer.MIN_VALUE);
        for (int num : nums) {
            if (num == maxi || num == mini)
                continue;
            int bucketIdx = (num - mini) / bucketSize;
            bucketMin[bucketIdx] = Math.min(bucketMin[bucketIdx], num);
            bucketMax[bucketIdx] = Math.max(bucketMax[bucketIdx], num);
        }
        int res = 0;
        int prevMax = mini;
        for (int i = 0; i < bucketLength; i++) {
            if (bucketMin[i] == Integer.MAX_VALUE)
                continue;
            res = Math.max(res, bucketMin[i] - prevMax);
            prevMax = bucketMax[i];
        }
        res = Math.max(res, maxi - prevMax);
        return res;
    }
}